src/Middleware/TrustedHostsNetworkResolver.php

```diff
@@ -0,0 +1,510 @@
+                 ] as $key) {
+            $this->checkStringArrayType($data[$key], $key);
+        }
+        foreach ($data[self::DATA_KEY_HOSTS] as $host) {
+            $host = str_replace('*', 'wildcard', $host);        // wildcard is allowed in host
+            if (filter_var($host, FILTER_VALIDATE_DOMAIN) === false) {
+                throw new \InvalidArgumentException("'$host' host is not a domain and not an IP address");
+            }
+        }
+        $new->trustedHosts[] = $data;
+        return $new;
+    }
+
+    private function checkStringArrayType(array $array, string $field): void
+    {
+        foreach ($array as $item) {
+            if (!is_string($item)) {
+                throw new \InvalidArgumentException("$field must be string type");
+            }
+            if (trim($item) === '') {
+                throw new \InvalidArgumentException("$field cannot be empty strings");
+            }
+        }
+    }
+
+    /**
+     * @return static
+     */
+    public function withoutTrustedHosts()
+    {
+        $new = clone $this;
+        $new->trustedHosts = [];
+        return $new;
+    }
+
+    /**
+     * Request's attribute name to which trusted path data is added.
+     *
+     * The list starts with the server and the last item is the client itself.
+     *
+     * @return static
+     * @see getElementsByRfc7239
+     */
+    public function withAttributeIps(?string $attribute)
+    {
+        if ($attribute === '') {
+            throw new \RuntimeException('Attribute should not be empty');
+        }
+        $new = clone $this;
+        $new->attributeIps = $attribute;
+        return $new;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $actualHost = $request->getServerParams()['REMOTE_ADDR'];
+        $trustedHostData = null;
+        $trustedHeaders = [];
+        $ipValidator = $this->ipValidator ?? new Ip();
+        foreach ($this->trustedHosts as $data) {
+            // collect all trusted headers
+            $trustedHeaders = array_merge($trustedHeaders, $data[self::DATA_KEY_TRUSTED_HEADERS]);
+            if ($trustedHostData !== null) {
+                // trusted hosts already found
+                continue;
+            }
+            if ($this->isValidHost($actualHost, $data[self::DATA_KEY_HOSTS], $ipValidator)) {
+                $trustedHostData = $data;
+            }
+        }
+        $untrustedHeaders = array_diff($trustedHeaders, $trustedHostData[self::DATA_KEY_TRUSTED_HEADERS] ?? []);
+        $request = $this->removeHeaders($request, $untrustedHeaders);
+        if ($trustedHostData === null) {
+            // No trusted host at all.
+            if ($this->notTrustedBranch !== null) {
+                return $this->notTrustedBranch->process($request, $handler);
+            }
+            $response = $this->responseFactory->createResponse(412);
```

Comment:
@samdark `403` or any?

Reply:
I don't think it's a good idea to throw exception here. We're not trying to deny access based on IP, right? Setting `requestClientIp` to null sounds more appropriate.

Reply:
should I remove this part as well?
```php
            if ($this->notTrustedBranch !== null) {
                return $this->notTrustedBranch->process($request, $handler);
            }
```

Reply:
If that's ability to set custom error handler then yes.